lock-api/src/test/java/com/palantir/lock/client/ProfilingTimelockServiceTest.java

```diff
@@ -0,0 +1,236 @@
+        assertThat(profilingTimelockService.getFreshTimestamp()).isEqualTo(42L);
+    }
+
+    @Test
+    public void logsIfFailedOperationsAreSlow() {
+        RuntimeException exception = new RuntimeException("kaboom");
+        when(profilingTimelockService.getFreshTimestamp()).thenThrow(exception);
+
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        allowLogging();
+        assertThatThrownBy(profilingTimelockService::getFreshTimestamp).isEqualTo(exception);
+
+        verify(delegate).getFreshTimestamp();
+        verifyLoggerInvokedWithSpecificProfile("getFreshTimestamp", LONG_DURATION);
+    }
+
+    @Test
+    public void doesNotLogIfLockIsSlow() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        allowLogging();
+        profilingTimelockService.lock(LockRequest.of(ImmutableSet.of(StringLockDescriptor.of("exclusive")), 1234));
+
+        verifyLoggerNeverInvoked();
+    }
+
+    @Test
+    public void doesNotLogIfWaitForLocksIsSlow() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        allowLogging();
+        profilingTimelockService.waitForLocks(WaitForLocksRequest.of(
+                ImmutableSet.of(StringLockDescriptor.of("combination")), 123));
+
+        verifyLoggerNeverInvoked();
+    }
+
+    @Test
+    public void logsSlowestOperationIfMultipleOperationsExceedTheSlowThresholdIfFirst() {
+        makeOperationsTakeSpecifiedDuration(TWO_CENTURIES);
+        profilingTimelockService.getFreshTimestamp();
+
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        allowLogging();
+        profilingTimelockService.startIdentifiedAtlasDbTransaction();
+
+        verifyLoggerInvokedWithSpecificProfile("getFreshTimestamp", TWO_CENTURIES);
+    }
+
+    @Test
+    public void logsSlowestOperationIfMultipleOperationsExceedTheSlowThresholdIfNotFirst() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        profilingTimelockService.getFreshTimestamp();
+
+        makeOperationsTakeSpecifiedDuration(TWO_CENTURIES);
+        allowLogging();
+        profilingTimelockService.startIdentifiedAtlasDbTransaction();
+
+        verifyLoggerInvokedWithSpecificProfile("startIdentifiedAtlasDbTransaction", TWO_CENTURIES);
+    }
+
+    @Test
+    public void logsTheSlowestNonBannedOperationAboveThreshold() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        profilingTimelockService.getFreshTimestamp();
+
+        makeOperationsTakeSpecifiedDuration(TWO_CENTURIES);
+        allowLogging();
+        profilingTimelockService.lock(LockRequest.of(ImmutableSet.of(StringLockDescriptor.of("mortice")), 2366));
+
+        verifyLoggerInvokedWithSpecificProfile("getFreshTimestamp", LONG_DURATION);
+    }
+
+    @Test
+    public void logsOnlyOnceEvenIfManyRequestsAreSlow() {
+        makeOperationsTakeSpecifiedDuration(LONG_DURATION);
+        for (int i = 0; i < 100; i++) {
+            profilingTimelockService.lockImmutableTimestamp();
```

Comment:
what about different calls, but same duration? do we care about that?

Reply:
We could do an 'any one of' check. I didn't intend there to be any guarantees as to which one would be returned in that case